extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunner.java

```diff
@@ -0,0 +1,571 @@
+ * One additional feature of this class is that kubernetes is the source of truth, so if you launch a task
+ * shutdown druid, bring up druid, the task will keep running and the state will be updated when the cluster
+ * comes back.  Thus while no tasks are technically restorable, all tasks once launched will run in isolation to the
+ * extent possible without requiring the overlord consistently up during their lifetime.
+ */
+
+public class KubernetesTaskRunner implements TaskLogStreamer, TaskRunner
+{
+
+  private static final EmittingLogger log = new EmittingLogger(KubernetesTaskRunner.class);
+  private final CopyOnWriteArrayList<Pair<TaskRunnerListener, Executor>> listeners = new CopyOnWriteArrayList<>();
+
+  // to cleanup old jobs that might not have been deleted.
+  private final ScheduledExecutorService cleanupExecutor;
+
+  protected final ConcurrentHashMap<String, K8sWorkItem> tasks = new ConcurrentHashMap<>();
+  private final TaskConfig taskConfig;
+  private final StartupLoggingConfig startupLoggingConfig;
+  private final TaskAdapter<Pod, Job> adapter;
+
+  private final KubernetesTaskRunnerConfig k8sConfig;
+  private final TaskQueueConfig taskQueueConfig;
+  private final TaskLogPusher taskLogPusher;
+  private final ListeningExecutorService exec;
+  private final KubernetesPeonClient client;
+  private final DruidNode node;
+
+
+  public KubernetesTaskRunner(
+      TaskConfig taskConfig,
+      StartupLoggingConfig startupLoggingConfig,
+      TaskAdapter<Pod, Job> adapter,
+      KubernetesTaskRunnerConfig k8sConfig,
+      TaskQueueConfig taskQueueConfig,
+      TaskLogPusher taskLogPusher,
+      KubernetesPeonClient client,
+      DruidNode node
+  )
+  {
+    this.taskConfig = taskConfig;
+    this.startupLoggingConfig = startupLoggingConfig;
+    this.adapter = adapter;
+    this.k8sConfig = k8sConfig;
+    this.taskQueueConfig = taskQueueConfig;
+    this.taskLogPusher = taskLogPusher;
+    this.client = client;
+    this.node = node;
+    this.cleanupExecutor = Executors.newScheduledThreadPool(1);
+    this.exec = MoreExecutors.listeningDecorator(
+        Execs.multiThreaded(taskQueueConfig.getMaxSize(), "k8s-task-runner-%d")
+    );
+    Preconditions.checkArgument(
+        taskQueueConfig.getMaxSize() < Integer.MAX_VALUE,
+        "The task queue bounds how many concurrent k8s tasks you can have, set it to a resonable value"
+    );
+  }
+
+
+  @Override
+  public Optional<InputStream> streamTaskLog(String taskid, long offset)
+  {
+    Optional<InputStream> peonLogs = client.getPeonLogs(new K8sTaskId(taskid));
+    if (!peonLogs.isPresent()) {
+      return Optional.absent();
+    }
+    return Optional.of(peonLogs.get());
+  }
+
+  @Override
+  public ListenableFuture<TaskStatus> run(Task task)
+  {
+    synchronized (tasks) {
+      tasks.computeIfAbsent(
+          task.getId(), k -> new K8sWorkItem(
+              task,
+              exec.submit(() -> {
```

Comment:
Suggestion: the block that follows is a rather complex (and critical) job state machine. It would be a bit easier to read if it were packed in its own class so a reader can focus on just the controller bits without the surrounding clutter.